### PR TITLE
Improved ajax

### PR DESCRIPTION
**Pull request by pangratz**
[Rendered](https://github.com/pangratz/rfcs/blob/ember-data-improved-ajax/text/0000-ember-data-improved-ajax.md)

### CODE REVIEW

**Comment by nathanhammond** [comment]
If we're going through this effort is it worth trying to build something that decouples us from the underlying jQuery implementation? There is hopefully a future in which we don't ship that by default and it would be sad to require jQuery just for `$.ajax`.

See: https://github.com/ember-cli/ember-ajax

**Comment by pangratz** [comment]
@nathanhammond yep, I definitely agree. The hooks should be designed to be agnostic of the specific library/implementation which does the actual request. I forgot to mention that explicitly in the RFC...

**Comment by mkay581** [comment]
@nathanhammond [fetch](https://fetch.spec.whatwg.org/) maybe?

**Comment by tchak** [comment]
I think the hooks as described in this RFC have no dependency on `$.ajax`. They are some examples based on `$.ajax` usage, men we could remove them, or add more examples with `fetch` for example.

One suggestion I have is that we are passing to `makeRequest()` a hash of options. The hash looks like this:
```
{
  url: 'https://api.com/posts',
  method: 'POST',
  headers: {},
  body: '{ post: {} }'
}
```

There is a spec for a [`Request` object](https://developer.mozilla.org/en-US/docs/Web/API/Request) (as part of `fetch`). I am not saying we should use it directly, but, I think we should model a `Request` object in `ember-data` after it. `DS.Request`? I would also expect both be compatible and so once we support fetch, it would be possible to pass in native `Request` object.

File: text/0000-ember-data-improved-ajax.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+calls the new public hooks.
+
+``` js
+_requestFor(params) {
+  let method  = this.methodForRequest(params);
+  let url     = this.urlForRequest(params);
+  let headers = this.headersForRequest(params);
+  let body    = this.bodyForRequest(params);
+
+  return RSVP.hash({ method, url, headers, body });
```
Comment (bmac) [inline text/0000-ember-data-improved-ajax.md]: Should `_requestFor` be made a public method in case the user wants to add some form of config other then method, url, headers or body?

File: text/0000-ember-data-improved-ajax.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+describe the request which should be made.
+
+``` js
+/**
+  @method
+  @public
+  @param request Object containing all properties for the request which should be made
+  @return Promise
+ */
+makeRequest(request) {
```
Comment (bmac) [inline text/0000-ember-data-improved-ajax.md]: I think we should enumerate all of the properties Ember Data expects/supports on the `request` object.

Comment (pangratz) [inline text/0000-ember-data-improved-ajax.md]: Good point!

Comment (bmac) [inline text/0000-ember-data-improved-ajax.md]: Based on my reading of this rfc it looks like it supports  `url`, `method`, `headers` and `body`.  With the exception of `url` all of these properties are shared by the 2nd argument to `fetch` (See: https://fetch.spec.whatwg.org/#requestinit). Which may lead a user to believe it supports all the properties that `fetch` supports.

This leads me to ask 2 questions.
1. How much of the RequestInit api should Ember Data implement?
2. Should the `url` property be pulled out of the request object and `makeRequest` API be change to align more closely with the `fetch` api? `makeRequest(url, requestInit)`

**Comment by bmac** [comment]
@tchak I'm not sold on a `DS.Request` class. What is the advantage of that wrapping the request state in a class instance vs just having users return a plain object with the same properties?

**Comment by bmac** [comment]
@pangratz in response to you unresolved questions section at the end of the rfc.
> currently requestType is used to indicate the specific adapter operation, but that name might not be to expressive. Should this be renamed to the more verbose, like adapterOperation?

I'm in favor of renaming `requestType` to `adapterOperation`.

> should queryForRequest be named queryParamsForRequest?

Again I like this change but I don't feel very strongly about it.

> should there be more hooks within urlForRequest, as outlined in the [Syntax](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax) section on the URLs wiki page?

I do not think we need any more hooks inside `urlForRequest` other then what is described in this rfc. I believe the current hooks satisfy most of not all of the requests for customization points in the url that we've seen in the Ember Data repo over the years. I suspect if we added more hooks they would be unused in 99% of cases and the remaining cases users can still override `urlForRequest` to generate the url they need.

> - the currently proposed API doesn't allow to only use public API to customize the json-api-adapter, though the current implementation could be changed to set the Content-Type header instead (which would be more agnostic anyway). This might be an indicator that the proposed API still needs some adjustments so json-api adapter uses only public API.
> - what hooks are missing?
>    - setting the dataType [#4357](https://github.com/emberjs/data/pull/4357)?

I was originally thinking `_requestFor` should be made public so users can return other config options that the [fetch api supports](https://fetch.spec.whatwg.org/#requestinit) besides `headers`, `method` and `body`. But that doesn't solve the case of special jquery specific configuration. 

2 possible approches:
1. Limit the ajax adapter to only allow configuring of jQuery ajax features that can be described using a fetch api. (For example contentType would use the Content-Type header but configuring `dataType` would be unsupported.) The rational being this would make it easier to move the rest adapter to a service worker some day since xhr isn't supported inside a service worker but fetch is.
2. Add a `jqueryAjaxOptions()` method and document that it will only be called when jQuery is being used for `makeRequest()` (Possibly difficult to detect).

File: text/0000-ember-data-improved-ajax.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+``` js
+urlForRequest(options) {
+  let path = this.pathForRequest(options);
+  let query = this.queryForRequest(options);
+
+  if (path.indexOf('?') !== -1) {
+    warning("use queryForRequest to specify query params");
+  } else if (query) {
+    let sortedQueryParams = this.sortQueryParams(query);
+    let queryParams = Ember.$.param(sortedQueryParams);
```
Comment (pangratz) [inline text/0000-ember-data-improved-ajax.md]: You mean additionally to `queryForRequest`?

Comment (pangratz) [inline text/0000-ember-data-improved-ajax.md]: Hmm, I see. Maybe `queryStringForRequest()`?

But then we would have 5 hooks which are involved getting the URL for a request:

- `urlForRequest`
  - `pathForRequest`
  - `queryForRequest`
  - `sortQueryParams`
  - `queryStringForRequest`

For now I am not convinced that an additional hook is necessary. Can you think of an use case where such a hook would come in handy?

Comment (bmac) [inline text/0000-ember-data-improved-ajax.md]: I've seen a few comments on slack recently were people want arrays or nested query params to be serialized differently from how `$.param` does it.

I wonder how widely used `sortQueryParams` is and I wonder if we could deprecate that and let users do that work in `queryForRequest` or `queryStringForRequest` if they need it.

Comment (pangratz) [inline text/0000-ember-data-improved-ajax.md]: Ah, arrays and nested queries are definitely a valid use case 👍 

Deprecating `sortQueryParams` in favor of `queryStringForRequest` sounds reasonable to me!

File: text/0000-ember-data-improved-ajax.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+  let headers = this.headersForRequest(params);
+  let body    = this.bodyForRequest(params);
+
+  return RSVP.hash({ method, url, headers, body });
+}
+```
+##### `methodForRequest`
+
+Invoked to get the HTTP method for the request. Should return one of `"GET"`,
+`"PUT"`, `"POST"` or `"DELETE"`.
```
Comment (woprandi) [inline text/0000-ember-data-improved-ajax.md]: Why not "PATCH" ?

**Comment by amiel** [comment]
This will help resolve some issues we've had with ember-data-url-templates :)

**Comment by cibernox** [comment]
This seems to have fallen idle. This would be much needed in order the build a reasonable compatibility layer to be able to swap jQuery by just raw fetch. 
At the moment adapters setup headers using the `beforeSend` function on ajax, which has no fetch counterpart and it is hard, if not plain impossible, to introspect.

Is there any open question or significant pushback from anyone that prevents this from advancing?

**Comment by denzo** [comment]
Same as above comment. It has been a while, it would be great to know where it's at or where it is on the timeline.

Will be happy to contribute to speed it up, if anything at all is needed.

The fact that https://emberjs.com/api/ember-data/2.16/classes/DS.RESTAdapter/methods/headersForRequest?anchor=headersForRequest is still in the docs but can not be used isn't ideal.

@pangratz @bmac @tchak

**Comment by tchak** [comment]
Her is the list of remarks and concerns so far on this RFC:
- make `requestFor` public
- the RFC dose not provide a way to customise `json-api` adapter with public methods
- `queryForRequest ` -> `queryParamsForRequest `
- deprecating `sortQueryParams` in favour of `queryStringForRequest`

"make things more standard":
- return `Request` object from `requestFor` (?)
- `makeRequest` to take `Request` (?)
- return a `URLSearchParams` from `queryParamsForRequest ` (?)

**Comment by wycats** [comment]
@emberjs/ember-data-contributors @igorT what do we need to do to make progress on these questions?

**Comment by sandstrom** [comment]
Regarding `headersForRequest`, I think it would be good if it returned a promise that resolved to a hash of headers (i.e. any consumer must expect either an object or a promise).

It's common to append authentication/session tokens in headers, and they may be stored asynchronously (for example in IndexedDB).

**Comment by fivetanley** [comment]
> return a URLSearchParams from queryParamsForRequest

@tchak is this polyfillable? This isn't supported by Internet Explorer natively.

**Comment by rwjblue** [comment]
Where are we on this? Is this still a direction we would like to go?

**Comment by tchak** [comment]
@rwjblue I think we want to finish migration to `fetch` in `ember-data` before adding anything else to the network layer. For what it worth, I implemented this RFC (and other things) in an addon https://github.com/tchak/ember-fetch-adapter

**Comment by runspired** [comment]
Closing as we are definitely moving in a very different direction. Keep your eyes open for a new RFC :)
